### ISSUE TITLE
Allocation of languages to train/dev/test

### ISSUE THREAD

**Issue by emilymbender** [gdtreebank/gdtreebank#2]
If I've understood correctly, fi_ftb and fi are both treebanks for Finnish.  It is therefore inappropriate for one to be in dev and the other to be in test --- if people are using this split to test claims of language independence.

More broadly, I understand that the allocation of treebanks to train/dev/test was done 'randomly', but I'd like to suggest that doing it with reference to language families would be a better strategy. For example, Danish, Swedish and Norwegian are *extremely* similar languages.  The allocation should take note of this fact.

**Comment by jeisner** [comment]
See below for several reasons why this split is not always "inappropriate."

First, though, let's be clear that the GD treebank release does not specify any particular split.  So you are complaining not about the GD dataset itself, but rather about some experiments that we did in [Wang & Eisner (2016)](https://www.cs.jhu.edu/~jason/papers/#wang-eisner-2016) to argue that GD was a useful extension to UD.

(You seemed to say on Twitter that the dataset was "fundamentally flawed," but surely that goes too far.  GD is just an undifferentiated collection of treebanks, like UD.  Your concern is with how we _used_ the dataset, especially if it leads others to use it in the same way.)

Ok, now to your point:

I agree that experimental design is important when using either UD or GD to test generalization from one corpus, language, or family to another.  As you suggest, what constitutes an appropriate split of the languages will depend on the hypothesis being tested.  

I understand your argument that we should not have used a random split in our experiments, but should instead have partitioned the data in some other way to demonstrate generalization.  However, there are several subtle issues here.  Here are my initial thoughts.  I am happy to discuss further.

1. It would be too extreme to say that training sets and test sets must not overlap.  In the usual ML setting, it is perfectly fine if they have some _types_ in common.  For example, the Penn Treebank test set might easily contain duplicates of some sentences in the Penn Treebank training set.  This is okay as long as they are different _tokens_ of that sentence, generated independently.  

After all, the test set is simply a sample from some distribution.  It is needed because the usual experimental design uses the _sample_ mean of some quantity (e.g., parse accuracy) as an estimate of the _population_ mean under that distribution.  For this purpose, the sample must be random, so we cannot prevent it from coincidentally overlapping with the training set.

How does that apply to the present circumstance?  If Romance-language corpora -- or even Finnish corpora -- are common in the world, as they apparently are, then we would expect to find them not only in test collections but also in training collections.  This is helpful, because it means that when the world gives us a new Finnish corpus to parse, we will do well thanks to having trained on other Finnish corpora.  Our good population performance on the world's distribution will be reflected in the estimate from the test sample, which will also contain Finnish corpora.

2. The above is a justification of the random split strategy.  At this point, you may complain about domain mismatch.  The world actually provides not one but two distributions: high-resource corpora for which parse trees are available, and low-resource corpora for which they aren't.  Our training set must be from the former distribution, because that is the only way to train.  But we would like our test sample to be from the latter distribution, because those are the only languages on which real users will benefit from our methods.  

Our paper actually addressed this point.  For practical reasons, of course, we cannot evaluate on corpora whose parse trees aren't available.  But we tried to simulate the above design as best we could.  We took our training languages to be the "highest-resource" UD languages -- those with the largest corpora.  We then took our test set and our dev set randomly from among the remaining "medium-resource" languages, as a proxy for actual low-resource languages.  As a result, all the modern Romance languages fell in the training data, and our experiments therefore tested whether it was possible to generalize from them to other types of languages.  You probably approve of that much.

Surprisingly, it turned out that Finnish is popular enough among creators of medium-resource corpora that (different) Finnish corpora showed up in both the dev set and the test set.  Different Latin corpora showed up in train, dev, and test.  Again, this is the luck of the draw.  It does suggest that in future, we may be called on to run our system on additional Finnish or Latin corpora ...

3. Sometimes the different corpora of the same language really are rather different!  This may reflect differences in genre or in annotation conventions.  Dingquan tells me that he checked this at the time with a small experiment (which of course excluded test data), and it contributed to his decision to keep the same split for the 2017 paper, rather than grouping multiple treebanks of the same language together.

Specifically, in single-source transfer, he found that the la_itt Latin parser was only the 11th best out of 19 parsers at parsing the la_proiel Latin corpus, and conversely, the la_proiel Latin parser was only the 7th best out of 10 parsers at parsing the la_itt Latin corpus (it's far better to use a Portuguese parser).

Dingquan writes about the multiple Latin and Finnish treebanks: "I was thinking that the publisher of UD separated them on purpose because of different annotation schemes."  On the other hand, he combined the Czech treebanks into a single dataset because they were not separated: "The Czech data actually has three treebanks under the same annotation scheme but different domains - in this case, all of them were placed under the same folder."

Note that the UD folks are concerned with standardizing the treatment of multiple corpora of the same language, and perhaps also related languages.  They are currently discussing these tricky issues [here](https://github.com/UniversalDependencies/docs/issues/463).  Other things equal, we will try to follow their lead in the next release of GD.

4. As noted in 2., we primarily measured generalization from high-resource to medium-resource languages.  What other goals are possible?

One might also be interested in measuring generalization to wholly unseen language families (rather than generalization to unseen languages within possibly known families).  While we are naturally curious what the results would be, we did not attempt this, for three reasons:
* There are even fewer language families than languages.  If we took our IID examples to be families rather than individual languages, we would have too few examples to expect statistical significance.
* There is not much engineering motivation for this case.  The usual engineering setting is to generalize to a low-resource language from the most closely related high-resource language, which is usually in the same family.
* We did not see much motivation for this case from human language learning, either.  The human language acquisition device (LAD) does not observe training languages; it is presumably designed to be able to handle any humanly possible unseen language.  In our followup paper on typology prediction, [Wang & Eisner (2017)](https://www.cs.jhu.edu/~jason/papers/#wang-eisner-2017), we tried to construct a kind of weak artificial LAD by training on some high- and medium-resource languages.  We wanted these training languages to reflect the actual diversity of possible human languages, including languages of the sort that we would test on.  

But we do agree with you that since our typology prediction system was trained on some languages that were actually related to (or identical to!) the languages that we would test on, it is more tailored to the task of analyzing our test languages than a human LAD could be.  After all, the human LAD's evolution was not influenced by any surviving language families.  One of our motivations in the 2017 paper was to show that an LAD-like device was possible -- and as you suggest, this proof of principle would indeed have been stronger if we had split by family.  Our other motivation was an engineering motivation, and from that perspective, our experimental design still seems okay as discussed in 1.-2. above.

5. In the 2016 paper, we were _not_ primarily trying to show that transfer was _possible_ from high-resource to medium-resource languages.   Rather, we were trying to show that this transfer was _improved_ if we augmented the high-resource UD languages with synthetic GD languages derived from them.

For this goal, I think having Latin corpora in both train and test must have harmed our argument, not helped it.  Under UD, the Latin test language could in principle have identified the Latin training language in the 10-language training collection as a good single source.[*]  Under GD, however, the Latin test language would presumably have had a harder time finding the Latin training language among the 1210 training languages.  On this argument, the reuse of Latin would have given UD a larger advantage -- making it harder, not easier, for us to show that GD training outperformed UD training.

[*] Although in practice it was not a good single source, as noted in 3. above, and would not / should not have been chosen.

**Comment by emilymbender** [comment]
The canonical citation for this dataset is Wang & Eisner 2016, which includes a train/dev/test split. Any future work that approaches the same questions as Wang & Eisner 2016 will be held accountable to matching that same split.  Furthermore, anyone picking up the dataset for other crosslinguistic questions that involve testing on real rather than synthetic languages will also quite naturally follow that split. 

I think you're splitting hairs and I maintain that this is a fundamental flaw in the dataset --- or in the dataset as presented, if you prefer.  Fortunately, it is one that you can address, by rerunning any experiments you have published with these data with a more appropriate data split, and documenting that on the page for the dataset so that others can find it and use it.

To your further points: 

(1) The analogy to parsing, where some sentence types recur, is invalid. In Wang & Eisner 2016, you write "As a result, it is challenging to develop systems that will discover structure in new languages in the same way that an image segmentation method, for example, will discover structure in new images. [...] Some such projects evaluate their methods on new sentences of the same languages that were used to develop the methods in the first place—which leaves one worried that the methods may be inadvertently tuned to the development languages and may not be able to discover correct structure in other languages." (p.491) and "If we could make labeled languages as plentiful as labeled images, then we could treat linguistic structure dis- covery as a problem of supervised prediction—one that need not succeed on all formal languages, but which should generalize at least to the domain of possible human languages." (p.502) 

And there I think you have it right. The scientific claim being tested is not whether your system can genralize to a random sample of existing treebanks. It is whether your system can generalize to an unseen language.

(3) Again, my concern is not whether you merge the Latin treebanks into one or not.  It's whether you keep the held-out languages held-out.

(4) That is not how you frame it in the paper (see quotes above).

"There is not much engineering motivation for this case. The usual engineering setting is to generalize to a low-resource language from the most closely related high-resource language, which is usually in the same family." --- If that is your reasoning, then you should have used engineering framing in the paper, rather than scientific framing.

(5) So rerun the experiments properly and share the results.  Also, note that it's not just Latin but also Finnish that is affected.  Latin appears to be represented in all three of train/dev/test.  Finnish is in dev & test.  You got lucky with Greek -- both Greek treebanks are in dev.  

More generally, if you get in the practice of listing actual language names (rather than treebank identifiers) these kinds of things will be immediately obvious to you and anyone else who picks up this data resource as you do your work, hopefully preventing such poor data handling practices in the future.
